src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/LanguageServerProjectLoader.cs

```diff
@@ -0,0 +1,276 @@
+            // language in-process.
+            var projectLanguage = loadedProjectInfos.FirstOrDefault()?.Language;
+            if (projectLanguage != null && ProjectFactory.Workspace.Services.GetLanguageService<ICommandLineParserService>(projectLanguage) == null)
+            {
+                return false;
+            }
+
+            var existingProjects = _loadedProjects.GetOrAdd(projectPath, static _ => []);
+
+            Dictionary<ProjectFileInfo, ProjectLoadTelemetryReporter.TelemetryInfo> telemetryInfos = [];
+            var needsRestore = false;
+
+            foreach (var loadedProjectInfo in loadedProjectInfos)
+            {
+                // If we already have the project with this same target framework, just update it
+                var existingProject = existingProjects.Find(p => p.GetTargetFramework() == loadedProjectInfo.TargetFramework);
+                bool targetNeedsRestore;
+                ProjectLoadTelemetryReporter.TelemetryInfo targetTelemetryInfo;
+
+                if (existingProject != null)
+                {
+                    (targetTelemetryInfo, targetNeedsRestore) = await existingProject.UpdateWithNewProjectInfoAsync(loadedProjectInfo, _logger);
+                }
+                else
+                {
+                    var projectSystemName = $"{projectPath} (${loadedProjectInfo.TargetFramework})";
+                    var projectCreationInfo = new ProjectSystemProjectCreationInfo
+                    {
+                        AssemblyName = projectSystemName,
+                        FilePath = projectPath,
+                        CompilationOutputAssemblyFilePath = loadedProjectInfo.IntermediateOutputFilePath
+                    };
+
+                    var projectSystemProject = await ProjectFactory.CreateAndAddToWorkspaceAsync(
+                        projectSystemName,
+                        loadedProjectInfo.Language,
+                        projectCreationInfo,
+                        _projectSystemHostInfo);
+
+                    var loadedProject = new LoadedProject(projectSystemProject, ProjectFactory.Workspace.Services.SolutionServices, _fileChangeWatcher, _targetFrameworkManager);
+                    loadedProject.NeedsReload += (_, _) => ProjectsToLoadAndReload.AddWork(projectToLoad with { ReportTelemetry = false });
+                    existingProjects.Add(loadedProject);
+
+                    (targetTelemetryInfo, targetNeedsRestore) = await loadedProject.UpdateWithNewProjectInfoAsync(loadedProjectInfo, _logger);
+
+                    needsRestore |= targetNeedsRestore;
+                    telemetryInfos[loadedProjectInfo] = targetTelemetryInfo with { IsSdkStyle = preferredBuildHostKind == BuildHostProcessKind.NetCore };
+                }
+            }
+
+            if (projectToLoad.ReportTelemetry)
+            {
+                await _projectLoadTelemetryReporter.ReportProjectLoadTelemetryAsync(telemetryInfos, projectToLoad, cancellationToken);
```

Comment:
No reason to address this in this PR, but @dibarbet one thing we'll have to think about is how we want to handle telemetry for single-file apps.

Reply:
Is there something specific we want to do?  We should probably add a property indicating it is a single file app (if there is no other way to tell from the telemetry report already).

Not sure how much telemetry we'll get though since it requires devkit be installed

Reply:
We do pass `/features:FileBasedProgram` to compiler, would that be something telemetry can see?

I would expect that many users using FBP would have devkit installed even if the project system used for their FBPs comes from here.

Reply:
> We do pass /features:FileBasedProgram to compiler, would that be something telemetry can see?

Only if we were to thread that through. Which might be all we want to make sure we do for now.

Reply:
But @dibarbet do we have reports that report 'time to first load' or something we need to update to ensure we don't accidentally count a file-based program as the 'first load'?

Reply:
> But @dibarbet do we have reports that report 'time to first load' or something we need to update to ensure we don't accidentally count a file-based program as the 'first load'?

This is mainly for tracking usage of .NET (and various workloads - for example we report what TFMs the project targets).  Should .NET run programs be tracked as part of that?  I assume so, but that might be a Damian question.

It isn't used for performance information currently.